### PR TITLE
2.0

### PR DESCRIPTION
**Pull request by kovrus**

### CODE REVIEW

File: docs/index.txt [hunk 1]
```diff
@@ -131,48 +129,23 @@ Properties can be specified when connecting to Crate using the JDBC driver::
...
-
-    .. note:: 3rd party tool ususally quote the identifiers obtained by the
-              ``getColumns()`` method. Quoting needs to be disabled when this
-              property is set to ``true``. This however limits the naming of
-              columns, because they must only contain lower case ASCII letters.
+    If you set this property to ``true`` the driver will be compliant with the JDBC
+    specification. It will then raise an unsupported operation exception if you
+    call an operation which is not supported, like auto-commit or setting a
+    savepoint.  The default is ``false``, so the Crate JDBC driver ignores the use
+    of unspported operations, which makes it compatible with most 3rd party
```
Comment (matriv) [inline docs/index.txt]: unsupported (missing **u**)

File: docs/index.txt [hunk 2]
```diff
@@ -252,6 +259,18 @@ there are elements in the array.
     arrayFieldResultSet.first();
     String firstValue = arrayFieldResultSet.getString("array_field");
 
+When inserting arrays into a prepared statement, you must convert your java array
```
Comment (matriv) [inline docs/index.txt]: ...arrays **using** a prepared...

File: docs/index.txt [hunk 3]
```diff
@@ -252,6 +259,18 @@ there are elements in the array.
...
+to a ``java.sql.Array`` through use of ``createArrayOf()``. This function takes
```
Comment (matriv) [inline docs/index.txt]: with the use of or by using

File: docs/index.txt [hunk 4]
```diff
@@ -214,20 +221,20 @@ for some crate types, for some it is not. This table shows how the crate types
...
+   "short", "int2", "SMALLINT", "``getShort``"
+   "integer", "int4", "INTEGER", "``getInteger``"
+   "long", "int8", "BIGINT", "``getLong``"
+   "float", "float4", "REAL", "``getFloat``"
+   "double", "float8", "DOUBLE", "``getDouble``"
+   "string", "varchar", "VARCHAR", "``getString``"
+   "timestamp", "timestamp", "TIMESTAMP", "``getDate``, ``getTime``, ``getTimestamp``"
+   "ip", "varchar", "VARCHAR", "``getString``"
+   "array", "", "ARRAY", "``getArray``"
+   "object", "json", "JAVA_OBJECT", "``getObject``"
```
Comment (matriv) [inline docs/index.txt]: Add geo_point, geo_shape & json

File: docs/index.txt [hunk 5]
```diff
@@ -181,29 +154,63 @@ Compatibility
...
+* the ``ResultSet`` is read only (``TYPE_FORWARD_ONLY``, ``CONCUR_READ_ONLY``),
+  so changes to a ``ResultSet`` are not supported. Though DDL and DML statements
+  are supported using the ``Statement`` and ``PreparedStatement`` interfaces.
+
+Server Version
+..............
+
+When the JDBC driver establishes a connection to Crate the server sends a messsage
+with its version to the client, so the client can behave according to the server
+version. Since the Crate JDBC driver builds upon the `PostgreSQL JDBC Driver`_
```
Comment (matriv) [inline docs/index.txt]: I would use "encapsulates" instead of builds upon

File: docs/index.txt [hunk 6]
```diff
@@ -181,29 +154,63 @@ Compatibility
...
+the Crate server needs to send a version that is usable by the client. In the
+case of Crate JDCB this is version ``9.5.0``.
+
+**This also means that some methods in ``DatabaseMetaData`` and ``Connection``
+do return this version instead of the real Crate server version!**
```
Comment (matriv) [inline docs/index.txt]: remove **do**

File: docs/index.txt [hunk 7]
```diff
@@ -181,29 +154,63 @@ Compatibility
 JDBC
 ----
 
-This JDBC driver complies to the `JDBC 4.1`_ standard as good as possible and reasonable for
-a Crate driver. The following mandatory features are not supported:
+This JDBC driver complies to the `JDBC 4.1`_ standard as good as possible and
```
Comment (matriv) [inline docs/index.txt]: should be as well as possible but don't like the whole phrase...
Maybe: This JDBC driver follows the JDBC 4.1 standard as it uses PGSQL driver as the underlying connector
but certain features are not fully compatible with Crate

File: docs/index.txt [hunk 8]
```diff
@@ -131,48 +129,23 @@ Properties can be specified when connecting to Crate using the JDBC driver::
...
-    By setting the property to ``false`` only the root column (the object
-    field name) itself is returned.
-
-    .. note:: 3rd party tool ususally quote the identifiers obtained by the
-              ``getColumns()`` method. Quoting needs to be disabled when this
-              property is set to ``true``. This however limits the naming of
-              columns, because they must only contain lower case ASCII letters.
+    If you set this property to ``true`` the driver will be compliant with the JDBC
+    specification. It will then raise an unsupported operation exception if you
+    call an operation which is not supported, like auto-commit or setting a
```
Comment (matriv) [inline docs/index.txt]: Would use 3rd person:
It will raise an Exception if an unsupported operation is called.

File: docs/index.txt [hunk 9]
```diff
@@ -131,48 +129,23 @@ Properties can be specified when connecting to Crate using the JDBC driver::
...
+    savepoint.  The default is ``false``, so the Crate JDBC driver ignores the use
```
Comment (matriv) [inline docs/index.txt]: "the calls" instead of "the use"

File: docs/index.txt [hunk 10]
```diff
@@ -77,44 +76,43 @@ JDBC Driver Class
...
-This will implicitly call ``setSchema`` on the ``Connection`` class. So the
-following two statements are equivalent::
+    crate://host1.example.com:5432,host2.example.com:5432/
 
-    Connection conn = DriverManager.getConnection("crate://localhost:4300/foo");
 
-Is the same as::
+To specify a schema, the ``setSchema`` method must be explicitly called on
+the connection. If no schema is provided, the ``doc`` default schema will be
+used::
```
Comment (matriv) [inline docs/index.txt]: .. code-block:: java

File: docs/index.txt [hunk 11]
```diff
@@ -131,48 +129,23 @@ Properties can be specified when connecting to Crate using the JDBC driver::
 
     Properties properties = new Properties();
     properties.put(<key>, <value>);
-    Connection conn = DriverManager.getConnection("crate://localhost:4300", properties);
+    Connection conn = DriverManager.getConnection("crate://localhost:5432/", properties);
 
 In addition connection properties can be passed via the JDBC URL::
```
Comment (matriv) [inline docs/index.txt]: .. code-block:: java

File: docs/index.txt [hunk 12]
```diff
@@ -5,70 +5,69 @@ Crate JDBC Driver
...
 interfaces to SQL-compliant databases.
 
-Crate provides a type 4 JDBC driver. The driver is written in Pure Java, and
-communicates in the database system's own network protocol. Because of this,
-the driver is platform independent.
+Crate provides a `type 4 JDBC driver`_. The driver is written in pure Java, and
+communicates with the database using the `PostgreSQL Wire Protocol`_.
 
-.. note:: This JDBC driver requires Crate 0.38.0 or newer!
+Because of this, all versions of the JDBC driver >= ``2.0`` require Crate ``0.57.0``
```
Comment (matriv) [inline docs/index.txt]: "Therefore" instead of "because of this"

File: README.rst [hunk 13]
```diff
@@ -31,19 +31,22 @@ Crate JDBC driver   Crate
 ``1.11.X``          ``0.54.X``
 ``1.12.X``          ``0.54.X``
 ``1.13.X``          ``0.55.X``
+``2.0.X``           ``0.57.X``
 =================== ==========
 
 Currently we don't provide the testing version of the Crate JDBC driver.
 Therefore, using the driver with the latest testing releases of the Crate
 might not be possible due compatibility issues. As a workaround, you can build
```
Comment (mfussenegger) [inline README.rst]: Starting with 2.X the compatibility issues should be gone - that's one of the major reason we're doing this after all

File: README.rst [hunk 14]
```diff
@@ -59,11 +62,27 @@ and build a JAR including all dependencies::
...
+   ./gradlew shadowJar
 
 Afterwards a JAR file of the current version exists under ``build/lib``.
 
 
+Version 2.x
+===========
+
+Version 2 of the Crate JDBC driver is using the `PostgreSQL Wire Protocol`_ and
+is currently only intended for testing purposes.
```
Comment (mfussenegger) [inline README.rst]: This is not true, starting with 0.57 this is the recommended thing.

File: driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java [hunk 15]
```diff
@@ -0,0 +1,589 @@
...
+        preparedStatement.execute();
+
+        preparedStatement =
+                connection.prepareStatement("insert into arrayTest (id, str_array, bool_array, byte_array, " +
+                        "short_array, integer_array, long_array, float_array, double_array, timestamp_array, " +
+                        "ip_array, obj_array) values " +
+                        "(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)");
+        preparedStatement.setInt(1, 1);
+        preparedStatement.setArray(2, connection.createArrayOf("string", new String[]{"a", "b", "c", "d"}));
+        preparedStatement.setArray(2, connection.createArrayOf("string", new String[]{"a", "b", "c", "d"}));
```
Comment (seut) [inline driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java]: is the twice binding here wanted

File: build.gradle [hunk 16]
```diff
@@ -62,28 +73,47 @@ task getVersion ( dependsOn: 'classes') << {
...
+    classifier ''
+    duplicatesStrategy 'fail'
     doLast {
         manifest {
             attributes("Implementation-Title": "Crate.IO JDBC Driver (Standalone)", "Implementation-Version": project.version)
         }
     }
+
+    relocate 'org.postgresql', 'io.crate.shade.org.postgresql'
+    relocate 'waffle', 'io.crate.shade.waffle'
```
Comment (mfussenegger) [inline build.gradle]: maybe add comments about what are pgjdbc deps and which are crate-jdbc deps

File: docs/index.txt [hunk 17]
```diff
@@ -181,29 +160,62 @@ Compatibility
...
+
 
-Though only an optional feature, it is still worth mentioning:
+**Though only an optional feature, it is still worth mentioning:**
+
+* the ``ResultSet`` is read only (``TYPE_FORWARD_ONLY``, ``CONCUR_READ_ONLY``),
+  so changes to a ``ResultSet`` are not supported. Though DDL and DML statements
+  are supported using the ``Statement`` and ``PreparedStatement`` interfaces.
+
+Server Version
```
Comment (mfussenegger) [inline docs/index.txt]: This is implementation detail, how is this relevant to a user?

File: docs/index.txt [hunk 18]
```diff
@@ -252,6 +266,18 @@ there are elements in the array.
...
 
+When inserting arrays using a prepared statement, you must convert your java array
+to a ``java.sql.Array`` by the use of ``createArrayOf()``. This function takes
+as its first argument a Crate type as described above and as its second the array
+you want to convert. You can then use ``setArray()`` to set this converted array
+using a prepared statement's argument.
+
+.. code-block:: java
+
+    PreparedStatement preparedStatement = connection.prepareStatement("insert into my_table " +
```
Comment (mfussenegger) [inline docs/index.txt]: In the crate project we have/had a java doctest layer to ensure examples are actually runnable. Might make sense to start using this here too?

File: driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java [hunk 19]
```diff
@@ -0,0 +1,589 @@
...
+
+    private static void insertIntoTable() throws SQLException {
+        Map<String, Object> objectField = new HashMap<String, Object>() {{
+            put("inner", "Zoon");
+        }};
+        PreparedStatement preparedStatement =
+            connection.prepareStatement("insert into test (id, string_field, boolean_field, byte_field, " +
+                "short_field, integer_field, long_field, float_field, double_field, object_field, " +
+                "timestamp_field, ip_field) values " +
+                "(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)");
```
Comment (seut) [inline driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java]: looks like that the complexity of this table (i mean all the columns and types) is only used at 1 test in that class.
would prefer a much simpler generic table used by most tests, and a complex one just created where it is needed (maybe same for the array table below)

File: driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java [hunk 20]
```diff
@@ -0,0 +1,589 @@
...
+            assertThat(resultSet.getLong(1), is(3L));
+        }
+    }
+
+    @Test
+    public void testExecuteBatchPreparedStatement() throws Exception {
+        try (Connection conn = DriverManager.getConnection(connectionStr)) {
+            PreparedStatement stmt = conn.prepareStatement("insert into test (id, string_field) values (?, ?)");
+            stmt.setInt(1, 2);
+            stmt.setString(2, "foo");
```
Comment (seut) [inline driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java]: is it somehow needed to use a string column in this test? if not remove it pls.

File: driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java [hunk 21]
```diff
@@ -0,0 +1,589 @@
...
+            assertThat(resultSet.getLong(1), is(4L));
+        }
+    }
+
+    @Test
+    public void testExecuteBatchPreparedStatementFailBulkTypes() throws Exception {
+        try (Connection conn = DriverManager.getConnection(connectionStr)) {
+            PreparedStatement stmt = conn.prepareStatement("insert into test (id, string_field) values (?, ?)");
+            stmt.setInt(1, 2);
+            stmt.setString(2, "foo");
```
Comment (seut) [inline driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java]: also here using a `id` column should be enough, cast failure could be done on that column as well afaik

File: driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java [hunk 22]
```diff
@@ -0,0 +1,589 @@
...
+            assertThat(resultSet.getLong(1), is(1L));
+        }
+    }
+
+    @Test
+    public void testExecuteBatchPreparedStatementFailOne() throws Exception {
+        try (Connection conn = DriverManager.getConnection(connectionStr)) {
+            PreparedStatement stmt = conn.prepareStatement("insert into test (id, string_field) values (?, ?)");
+            stmt.setInt(1, 2);
+            stmt.setString(2, "foo");
```
Comment (seut) [inline driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java]: also here, try to keep tests simple pls

File: driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java [hunk 23]
```diff
@@ -0,0 +1,589 @@
...
+            assertThat(resultSet.getLong(1), is(3L));
+        }
+    }
+
+    @Test
+    public void testExecuteBatchPreparedStatementFailSyntax() throws Exception {
+        try (Connection conn = DriverManager.getConnection(connectionStr)) {
+            PreparedStatement stmt = conn.prepareStatement("insert test (id, string_field) values (?, ?)");
+            stmt.setInt(1, 2);
+            stmt.setString(2, "foo");
```
Comment (seut) [inline driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java]: also here, string col is irrelevant

File: driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java [hunk 24]
```diff
@@ -0,0 +1,589 @@
...
+        }
+    }
+
+    @Test
+    @Ignore("getSchemas is not implemented")
+    public void testGetSchemas() throws Exception {
+        try (Connection conn = DriverManager.getConnection(connectionStr)) {
+            DatabaseMetaData metaData = conn.getMetaData();
+            ResultSet result = metaData.getSchemas();
+            result.next();
```
Comment (seut) [inline driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCConnectionTest.java]: pls test that getSchemas returns valid schemas only. testing for all implemented schemas is imho beyond of this test and will fail if other schemas are implemented by the server

File: driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCTypesTest.java [hunk 25]
```diff
@@ -0,0 +1,407 @@
...
+                "from information_schema.tables where schema_name " +
+                "not in ('pg_catalog', 'sys', 'information_schema', 'blob')");
+        while (rs.next()) {
+            connection.createStatement().execute(String.format("drop table if exists \"%s\".\"%s\"",
+                    rs.getString("schema_name"), rs.getString("table_name")));
+        }
+    }
+
+    public static void setUpTables() throws InterruptedException, SQLException {
+        connection.createStatement().execute("create table if not exists test (" +
```
Comment (seut) [inline driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCTypesTest.java]: looks similar to setUpTables of CrateJDBCConnectionTest, can we share?

File: driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCTypesTest.java [hunk 26]
```diff
@@ -0,0 +1,407 @@
...
+        rs.next();
+        return rs.getLong(1);
+    }
+
+    private static void insertIntoTable() throws SQLException {
+        Map<String, Object> objectField = new HashMap<String, Object>() {{
+            put("inner", "Zoon");
+        }};
+        PreparedStatement preparedStatement =
+                connection.prepareStatement("insert into test (id, string_field, boolean_field, byte_field, " +
```
Comment (seut) [inline driver/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCTypesTest.java]: looks similar to CrateJDBCConnectionTest.insertIntoTable, can we share?

File: README.rst [hunk 27]
```diff
@@ -70,10 +69,10 @@ Afterwards a JAR file of the current version exists under ``build/lib``.
 Version 2.x
 ===========
 
-Version 2 of the Crate JDBC driver is using the `PostgreSQL Wire Protocol`_ and
-is currently only intended for testing purposes.
-It is only available via `Jitpack`_ and won't be available on `Bintray`_ until
-Crate version ``0.57.x`` is released.
+Version 2 of the Crate JDBC driver is using the `PostgreSQL Wire Protocol`_.
+Starting from the Crate version ``0.57.0`` it is the recommended way to
+connect to Crate server. The Crate JDBC driver currently available via
```
Comment (mfussenegger) [inline README.rst]: > The Crate JDBC driver currently available via

there is an *is* missing
